qt/scientific_interfaces/Muon/ALCDataLoadingModel.cpp

```diff
@@ -0,0 +1,362 @@
+    QCoreApplication::processEvents();
+  }
+  if (!result.error().empty()) {
+    throw std::runtime_error(result.error());
+  }
+
+  // Set loading alg equal to alg
+  this->m_LoadingAlg = alg;
+
+  MatrixWorkspace_sptr tmp = alg->getProperty("OutputWorkspace");
+  IAlgorithm_sptr sortAlg = AlgorithmManager::Instance().create("SortXAxis");
+  sortAlg->setAlwaysStoreInADS(false); // Don't want workspaces in the ADS
+  sortAlg->setProperty("InputWorkspace", tmp);
+  sortAlg->setProperty("Ordering", "Ascending");
+  sortAlg->setProperty("OutputWorkspace", "__NotUsed__");
+  sortAlg->execute();
+
+  m_loadedData = sortAlg->getProperty("OutputWorkspace");
+
+  // If errors are properly caught, shouldn't happen
+  assert(m_loadedData);
+  // If subtract is not checked, only one spectrum,
+  // else four spectra
+  if (!subtractIsChecked) {
+    assert(m_loadedData->getNumberHistograms() == 1);
+  } else {
+    assert(m_loadedData->getNumberHistograms() == 4);
+  }
+}
+
+/**
+ * Remove the run number from a full file path
+ * @param file :: [input] full path which contains a run number
+ * @return An integer representation of the run number
+ */
+int ALCDataLoadingModel::extractRunNumber(const std::string &file) {
+  if (file.empty())
+    return -1;
+
+  auto returnVal = file;
+  // Strip beginning of path to just the run (e.g. MUSR00015189.nxs)
+  std::size_t found = returnVal.find_last_of("/\\");
+  returnVal = returnVal.substr(found + 1);
+
+  // Remove all non-digits
+  returnVal.erase(std::remove_if(returnVal.begin(), returnVal.end(), [](unsigned char c) { return !std::isdigit(c); }),
+                  returnVal.end());
+
+  // Return run number as int (removes leading 0's)
+  return std::stoi(returnVal);
+}
+
+std::string ALCDataLoadingModel::getPathFromFiles(const std::vector<std::string> &files) const {
+  if (files.empty())
+    return "";
+  const auto firstDirectory = files[0u].substr(0u, files[0u].find_last_of("/\\"));
+  // Lambda to compare directories from a path
+  const auto hasSameDirectory = [&firstDirectory](const auto &path) {
+    return path.substr(0u, path.find_last_of("/\\")) == firstDirectory;
+  };
+  const auto sameDirectory = std::all_of(files.cbegin(), files.cend(), hasSameDirectory);
+  return sameDirectory ? firstDirectory : "Multiple Directories";
+}
+
+/**
+ * If custom grouping is supplied, check all detector numbers are valid
+ * @returns :: True if grouping OK, false if bad
+ */
+bool ALCDataLoadingModel::checkCustomGrouping(const std::string &detGroupingType, const std::string &forwardGrouping,
+                                              const std::string &backwardGrouping) {
+  if (detGroupingType != "Custom") {
+    return true;
+  }
+
+  bool groupingOK = true;
+
+  auto detectors = Mantid::Kernel::Strings::parseRange(isCustomGroupingValid(forwardGrouping, groupingOK));
+  const auto backward = Mantid::Kernel::Strings::parseRange(isCustomGroupingValid(backwardGrouping, groupingOK));
```

Comment:
Am I right that the logic here means that the final value of the `groupingOK` variable is only actually being determined by the custom grouping of backward (i.e. by line 256)? The custom grouping of the detectors in the step above doesn't influence the final value of `groupingOK`?

I just wanted to double check that that makes sense for what this is trying to achieve? I'm not sure of the real world context of this, so sorry if I'm missing something obvious here!

Reply:
My goodness, that's really well spotted, I'll fix it

Reply:
In fairness, I think this was how the logic was originally, so if you prefer we could open an issue to fix the bug as a separate PR and get this one merged in?

Reply:
No, it's okay :relieved: